qemu/tests/block_hotplug_negative.py

```diff
@@ -0,0 +1,98 @@
+import logging
+import re
+import time
+
+from virttest import data_dir
+from virttest import storage
+from virttest import error_context
+from virttest.qemu_devices import qdevices
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Test hotplug of block devices.
+
+    1) Boot up guest without block device.
+    2) Hoplug block device with invalid blk params.
+    3) Unplug the drive
+    4) Hotplug the drive again
+
+    :param test:   QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
+    """
+    def find_image(image_name):
+        """
+        Find the path of the iamge.
+        """
+        image_params = params.object_params(image_name)
+        o = storage.get_image_filename(image_params, data_dir.get_data_dir())
+        return o
+
+    def drive_unplug_plug(drive, vm):
+        """
+        Unplug drive then replug it.
+        """
+        error_context.context("unplug the drive", logging.info)
+        drive.unplug(vm.monitor)
+        ver_out = drive.verify_hotplug("", vm.monitor)
+        time.sleep(5)
+        error_context.context("Hotplug the drive", logging.info)
+        drive.hotplug(vm.monitor)
+
+    img_list = params.get("images").split()
+    img_format_type = params.get("img_format_type", "qcow2")
+    pci_type = params.get("pci_type", "virtio-blk-pci")
+    blk_num = int(params.get("blk_num", 1))
+    add_block_device = True
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    error_context.context("Hotplug block device", logging.info)
+    for num in xrange(blk_num):
+        device = qdevices.QDevice(pci_type)
+        drive = qdevices.QRHDrive("block%d" % num)
+        drive.set_param("file", find_image(img_list[num + 1]))
+        drive.set_param("format", img_format_type)
+        drive_id = drive.get_param("id")
+        drive.hotplug(vm.monitor)
+        #add controller if needed
+        if params.get("need_controller", "no") == "yes":
+            controller_model = params.get("controller_model")
+            if params.get("bus_extra_params_%s" % img_list[num + 1]):
+                bus_extra_param = params["bus_extra_params_%s" % img_list[num + 1]]
+                match = re.search("(\w+)=(\w+)", bus_extra_param)
+                if match:
+                    qdevice_params = {match.group(1): match.group(2)}
+                    controller = qdevices.QDevice(controller_model, params=qdevice_params)
+            else:
+                controller = qdevices.QDevice(controller_model)
+            try:
+                controller.hotplug(vm.monitor)
+            except:
+                error_context.context("Failed to add controller with invalid params", logging.info)
+                drive_unplug_plug(drive, vm)
```

Comment:
why add unplug_plug to the except

Reply:
To check drive unplug_plug can work normally after failed to hotplug device.